docs/source/traitsui_user_manual/testing.rst

```diff
@@ -0,0 +1,633 @@
+
+    custom_registry = TargetRegistry()
+    custom_registry.register_handler(
+        target_class=ShinyButton,
+        interaction_class=ManyMouseClick,
+        handler=many_mouse_click,
+    )
+
+The signature of ``many_mouse_click`` is required by the |TargetRegistry.register_handler|
+method on |TargetRegistry|. By setting the ``target_class`` and
+``interaction_class``, we restrict the types of ``wrapper._target`` and
+``interaction`` received by ``many_mouse_click`` respectively.
+
+Finally, we can use this registry with the |UITester|::
+
+    tester = UITester(registries=[custom_registry])
+
+All the builtin testing support for TraitsUI editors are still present, but now
+this tester can perform the additional, custom user interaction.
+
+Add support for inspecting GUI states
+-------------------------------------
+
+The steps to extend |UIWrapper.inspect| is identical to those for extending
+|UIWrapper.perform| (see section above). The distinction between
+|UIWrapper.perform| and |UIWrapper.inspect| is merely in their returned
+values.
+
+In fact, following the steps in the above section, the new ``ManyMouseClick``
+can also be called via |UIWrapper.inspect|::
+
+    value = tester.inspect(ManyMouseClick())
+
+The returned value is the returned value from ``many_mouse_click``, which is
+``None``.
+
+Add support for locating a nested GUI element
+---------------------------------------------
+
+Support for |UIWrapper.locate| can be extended by registering additional
+location type and resolution logic via |TargetRegistry.register_solver| on
+a |TargetRegistry|.
+
+Suppose we have a custom UI editor that contains some buttons. The objective of
+a test is to click a specific button with a given label. We will therefore need
+to locate the button with the given label before a mouse click can be
+performed.
+
+The test code we wanted to achieve looks like this::
+
+    container = UITester().find_by_id(ui, "some_container")
+    button_wrapper = container.locate(NamedButton("OK"))
+
+where the targets are::
+
+    >>> container._target
+    <package.ui.qt.shiny_panel.ShinyPanel object at 0x7f940a3f10b8>
+    >>> button_wrapper._target
+    <package.ui.qt.shiny_button.ShinyButton object at 0x7fbcc3a63438>
+
+We can define the new ``NamedButton`` location type::
+
+    class NamedButton:
+        ''' Locator for locating a push button by label.'''
+        def __init__(self, label):
+            self.label = label
+
+Say ``ShinyPanel`` keeps track of the buttons with a dictionary called
+``_buttons`` where the names of the buttons are the keys of the dictionary.
+Then the logic to retrieving a button from a
+label can be written like this::
+
+    def get_button(wrapper, location):
+        """ Returns a ShinyButton from a UIWrapper wrapping ShinyPanel."""
+        # wrapper is an instance of UIWrapper
+        # location is an instance of NamedButton
+        return wrapper.target._buttons[location.label]
+
```

Comment:
should we mention that this signature is require by register_solver, like you did explicitly for register_handler in the example above?  It is probably clear by reading it, but it might not hurt to say it explicitly